File: easy/javascript/1071-greatest-common-divisor-of-string.js
/**
 * n = str1.len
 * m = str2.len
 * O(min(n,m) * (m+n))
 * @param {string} str1
 * @param {string} str2
 * @return {string}
 */
var gcdOfStrings = function (str1, str2) {
  let minStr = str1.length < str2.length ? str1 : str2;

  while (minStr.length > 0) {
    if (isGcd(str1, str2, minStr)) {
      break;
    }

    minStr = minStr.substring(0, minStr.length - 1);
  }

  return minStr;
};

function isGcd(str1, str2, gcd) {
  const len1 = str1.length;
  const len2 = str2.length;
  const gcdLen = gcd.length;

  if (len1 % gcdLen > 0 || len2 % gcdLen > 0) {
    return false;
  }

  const divideStr1 = gcd.repeat(len1 / gcdLen) === str1;
  const divideStr2 = gcd.repeat(len2 / gcdLen) === str2;

  return divideStr1 && divideStr2;
}

File: easy/go/1071-greatest-common-divisor-of-string.go

import "strings"

// O(min(n,m) * (m+n))
func gcdOfStrings(str1 string, str2 string) string {
	var minStr string

	if len(str1) > len(str2) {
		minStr = str2
	} else {
		minStr = str1
	}

	for len(minStr) > 0 {
		if isGcd(str1, str2, minStr) {
			break
		}

		minStr = minStr[:len(minStr) -1]
	}

	return minStr 
}

func isGcd(str1 string, str2 string, gcd string) bool{
	len1 := len(str1)
	len2 := len(str2)
	gcdLen := len(gcd)

	if len1 % gcdLen > 0 || len2 % gcdLen < 0 {
		return false
	}

	divideStr1 := strings.Repeat(gcd, len1 / gcdLen) == str1
	divideStr2 := strings.Repeat(gcd, len2 / gcdLen) == str2

	return divideStr1 && divideStr2
}